File: src/hyperhandler/risk/calculator.py
"""Risk calculations: ATR, position sizing, leverage selection."""

from decimal import ROUND_DOWN, Decimal

from hyperhandler.models.order import Position
from hyperhandler.models.risk import (
    CumulativeRiskResult,
    FundingEstimate,
    LeverageResult,
    PositionSizeResult,
    RejectReason,
    RiskReject,
    StopLossResult,
)
from hyperhandler.models.signal import SignalHorizon
from hyperhandler.risk.config import (
    ATR_SETTINGS,
    CORRELATION_MAP,
    HL_MAINTENANCE_MARGIN,
    MAX_CONFIDENCE_FACTOR,
    MIN_CONFIDENCE_FACTOR,
    HLConfig,
    RiskProfile,
)


class RiskCalculator:
    """Pure calculation functions for risk management."""

    def __init__(self, profile: RiskProfile, hl_config: HLConfig):
        self.profile = profile
        self.hl_config = hl_config

    def calculate_atr(
        self,
        candles: list[dict],
        period: int = 14,
    ) -> Decimal:
        """Calculate ATR using EMA method (pure Python, no numpy).

        Args:
            candles: HL candles with {o, h, l, c} fields
            period: ATR period (default 14)

        Returns:
            ATR value as Decimal

        Raises:
            ValueError: If insufficient candles
        """
        if len(candles) < 2:
            raise ValueError("Need at least 2 candles for ATR")

        true_ranges: list[Decimal] = []
        for i in range(1, len(candles)):
            high = Decimal(str(candles[i]["h"]))
            low = Decimal(str(candles[i]["l"]))
            prev_close = Decimal(str(candles[i - 1]["c"]))

            tr = max(
                high - low,
                abs(high - prev_close),
                abs(low - prev_close),
            )
            true_ranges.append(tr)

        if len(true_ranges) < period:
            # Not enough data for full EMA, use simple average
            return sum(true_ranges) / len(true_ranges)

        # EMA calculation
        alpha = Decimal("2") / (Decimal(str(period)) + Decimal("1"))
        ema = true_ranges[0]
        for tr in true_ranges[1:]:
            ema = alpha * tr + (Decimal("1") - alpha) * ema

        return ema

    def calculate_stop_loss(
        self,
        entry_price: Decimal,
        side: str,
        atr: Decimal,
        horizon: SignalHorizon,
    ) -> StopLossResult:
        """Calculate ATR-based stop-loss price.

        Args:
            entry_price: Entry price
            side: "long" or "short"
            atr: ATR value
            horizon: Signal horizon for multiplier selection

        Returns:
            StopLossResult with price and calculation details
        """
        settings = ATR_SETTINGS[horizon]
        multiplier = settings["multiplier"]

        stop_distance = atr * multiplier
        buffer = stop_distance * self.hl_config.slippage_buffer
        stop_with_buffer = stop_distance + buffer

        if side == "long":
            stop_price = entry_price - stop_with_buffer
        else:
            stop_price = entry_price + stop_with_buffer

        return StopLossResult(
            price=stop_price,
            distance=stop_with_buffer,
            distance_pct=stop_with_buffer / entry_price,
            atr_value=atr,
            atr_multiplier=multiplier,
        )

    def estimate_liquidation_price(
        self,
        entry_price: Decimal,
        leverage: int,
        side: str,
    ) -> Decimal:
        """Estimate liquidation price for a NEW position.

        Simplified formula (HL cross margin):
        liq_distance ≈ 1/leverage - maintenance_margin

        Args:
            entry_price: Expected entry price
            leverage: Selected leverage
            side: "long" or "short"

        Returns:
            Estimated liquidation price
        """
        liq_distance_pct = (Decimal("1") / Decimal(str(leverage))) - HL_MAINTENANCE_MARGIN
        liq_distance_pct = max(Decimal("0.01"), liq_distance_pct)  # Floor at 1%

        if side == "long":
            return entry_price * (Decimal("1") - liq_distance_pct)
        else:
            return entry_price * (Decimal("1") + liq_distance_pct)

    def validate_stop_vs_liquidation(
        self,
        stop_price: Decimal,
        liquidation_price: Decimal,
        entry_price: Decimal,
        side: str,
    ) -> bool:
        """Validate stop-loss is closer to entry than liquidation.

        Args:
            stop_price: Stop-loss price
            liquidation_price: Liquidation price (actual or estimated)
            entry_price: Entry price
            side: "long" or "short"

        Returns:
            True if valid, False if stop is beyond liquidation
        """
        if side == "long":
            # For long: stop must be ABOVE liquidation
            if stop_price <= liquidation_price:
                return False
        else:
            # For short: stop must be BELOW liquidation
            if stop_price >= liquidation_price:
                return False

        # Check safety buffer (stop should be at least 2% away from liq)
        liq_buffer = abs(stop_price - liquidation_price) / entry_price
        return liq_buffer >= self.hl_config.liq_safety_buffer

    def select_leverage(
        self,
        stop_distance_pct: Decimal,
        max_leverage_coin: int,
    ) -> LeverageResult:
        """Select leverage so liquidation is beyond stop-loss.

        Args:
            stop_distance_pct: Stop distance as % of entry
            max_leverage_coin: HL max leverage for this coin

        Returns:
            LeverageResult with selected leverage and reasoning
        """
        safety_factor = Decimal("1.5")

        if stop_distance_pct <= 0:
            max_safe = self.profile.max_leverage
        else:
            max_safe = int(Decimal("1") / (stop_distance_pct * safety_factor))
            max_safe = max(1, max_safe)

        leverage = min(
            max_safe,
            max_leverage_coin,
            self.profile.max_leverage,
        )

        reason = []
        if leverage == max_safe:
            reason.append("safe_for_stop")
        if leverage == max_leverage_coin:
            reason.append("coin_max")
        if leverage == self.profile.max_leverage:
            reason.append("config_max")

        return LeverageResult(
            leverage=leverage,
            max_safe=max_safe,
            max_coin=max_leverage_coin,
            max_config=self.profile.max_leverage,
            reason="+".join(reason) if reason else "default",
        )

    def select_leverage_for_stop(
        self,
        stop_price: Decimal,
        entry_price: Decimal,
        side: str,
        max_leverage_coin: int,
    ) -> LeverageResult:
        """Select leverage that makes the given stop valid.

        Used when ATR-based stop requires lower leverage.

        Args:
            stop_price: Stop-loss price
            entry_price: Entry price
            side: "long" or "short"
            max_leverage_coin: HL max leverage for this coin

        Returns:
            LeverageResult with adjusted leverage
        """
        stop_distance_pct = abs(entry_price - stop_price) / entry_price

        # Liquidation must be further than stop + safety buffer
        required_liq_distance = stop_distance_pct + self.hl_config.liq_safety_buffer

        # liq_distance ≈ 1/leverage - maintenance
        # leverage ≈ 1 / (liq_distance + maintenance)
        max_safe = int(Decimal("1") / (required_liq_distance + HL_MAINTENANCE_MARGIN))
        max_safe = max(1, max_safe)

        leverage = min(max_safe, max_leverage_coin, self.profile.max_leverage)

        return LeverageResult(
            leverage=leverage,
            max_safe=max_safe,
            max_coin=max_leverage_coin,
            max_config=self.profile.max_leverage,
            reason="adjusted_for_stop",
        )

    def calculate_position_size(
        self,
        account_value: Decimal,
        available_balance: Decimal,
        entry_price: Decimal,
        stop_price: Decimal,
        leverage: int,
        sz_decimals: int,
        confidence: float | None = None,
        risk_multiplier: Decimal = Decimal("1.0"),
        max_risk_amount: Decimal | None = None,
    ) -> PositionSizeResult | RiskReject:
        """Calculate position size based on risk budget.

        Args:
            account_value: Total account equity
            available_balance: Free margin
            entry_price: Entry price
            stop_price: Stop-loss price
            leverage: Selected leverage
            sz_decimals: Asset size decimals (for rounding)
            confidence: Signal confidence (0.0-1.0)
            risk_multiplier: Circuit breaker multiplier
            max_risk_amount: Budget constraint (max risk $)

        Returns:
            PositionSizeResult or RiskReject if position too small
        """
        # Clamp confidence factor
        if confidence is not None:
            confidence_factor = max(
                MIN_CONFIDENCE_FACTOR,
                min(MAX_CONFIDENCE_FACTOR, Decimal(str(confidence))),
            )
        else:
            confidence_factor = MAX_CONFIDENCE_FACTOR

        risk_pct = self.profile.risk_per_trade * confidence_factor * risk_multiplier
        risk_amount = account_value * risk_pct

        # Apply budget constraint
        if max_risk_amount is not None and risk_amount > max_risk_amount:
            risk_amount = max_risk_amount

        stop_distance = abs(entry_price - stop_price)
        if stop_distance == 0:
            return RiskReject(
                reason=RejectReason.ATR_UNAVAILABLE,
                details="Stop distance is zero",
                suggested_action="wait",
            )

        raw_size = risk_amount / stop_distance
        notional = raw_size * entry_price
        commission = notional * self.hl_config.taker_fee * Decimal("2")

        # Adjust for commission
        adjusted_risk = risk_amount - commission
        if adjusted_risk <= 0:
            return RiskReject(
                reason=RejectReason.POSITION_TOO_SMALL,
                details="Risk amount doesn't cover commission",
                suggested_action="wait",
            )

        adjusted_size = adjusted_risk / stop_distance

        # Check margin constraint
        margin_required = (adjusted_size * entry_price) / Decimal(str(leverage))
        if margin_required > available_balance:
            max_size = (available_balance * Decimal(str(leverage))) / entry_price
            adjusted_size = min(adjusted_size, max_size)
            margin_required = (adjusted_size * entry_price) / Decimal(str(leverage))

        # Round down to szDecimals
        adjusted_size = self._round_down(adjusted_size, sz_decimals)

        final_notional = adjusted_size * entry_price
        final_risk = adjusted_size * stop_distance
        final_commission = final_notional * self.hl_config.taker_fee * Decimal("2")

        # Check minimum order
        if final_notional < self.hl_config.min_order_value:
            return RiskReject(
                reason=RejectReason.POSITION_TOO_SMALL,
                details=f"Order ${final_notional:.2f} < min ${self.hl_config.min_order_value}",
                suggested_action="wait",
            )

        return PositionSizeResult(
            size=adjusted_size,
            notional=final_notional,
            margin_required=margin_required,
            risk_amount=final_risk,
            risk_pct=final_risk / account_value if account_value > 0 else Decimal("0"),
            commission_estimate=final_commission,
        )

    def calculate_cumulative_risk(
        self,
        open_positions: list[Position],
        new_risk_amount: Decimal,
        new_coin: str,
        account_value: Decimal,
    ) -> CumulativeRiskResult:
        """Calculate cumulative portfolio risk with correlation adjustment.

        Args:
            open_positions: List of open positions
            new_risk_amount: Risk amount of new position (0 for preview)
            new_coin: Coin of new position
            account_value: Total account equity

        Returns:
            CumulativeRiskResult with risk metrics
        """
        # Build correlation groups from positions
        groups: dict[str, list[str]] = {}
        for pos in open_positions:
            group = self._get_correlation_group(pos.coin)
            if group not in groups:
                groups[group] = []
            groups[group].append(pos.coin)

        # Add new coin if we're actually adding risk
        if new_risk_amount > 0:
            new_group = self._get_correlation_group(new_coin)
            if new_group not in groups:
                groups[new_group] = []
            if new_coin not in groups[new_group]:
                groups[new_group].append(new_coin)

        # Sum raw risk
        total_risk = sum(pos.risk_amount or Decimal("0") for pos in open_positions)

        # Calculate adjusted risk with correlation penalty
        adjusted_risk = Decimal("0")
        for group_name, coins in groups.items():
            group_positions = [p for p in open_positions if p.coin in coins]
            group_risk = sum(p.risk_amount or Decimal("0") for p in group_positions)

            n = len(coins)
            if n > 1:
                penalty = Decimal("1") + Decimal(str(n - 1)) * self.profile.correlation_factor
                group_risk = group_risk * penalty

            adjusted_risk += group_risk

        # Add cascade buffer for correlated groups
        cascade_buffer = Decimal("0")
        for group_name, coins in groups.items():
            if len(coins) > 1:
                group_positions = [p for p in open_positions if p.coin in coins]
                group_risk = sum(p.risk_amount or Decimal("0") for p in group_positions)
                cascade_buffer += group_risk * Decimal("0.1")

        total_adjusted = adjusted_risk + cascade_buffer + new_risk_amount
        max_risk = account_value * self.profile.max_cumulative_risk

        return CumulativeRiskResult(
            raw_risk=total_risk + new_risk_amount,
            adjusted_risk=total_adjusted,
            risk_pct=total_adjusted / account_value if account_value > 0 else Decimal("0"),
            available_budget=max(Decimal("0"), max_risk - adjusted_risk - cascade_buffer),
            within_limit=total_adjusted <= max_risk,
            correlation_groups=groups,
        )

    def estimate_funding_cost(
        self,
        size: Decimal,
        entry_price: Decimal,
        side: str,
        funding_rate: Decimal,
        risk_amount: Decimal,
        hold_hours: int = 24,
    ) -> FundingEstimate:
        """Estimate funding costs/income.

        Args:
            size: Position size
            entry_price: Entry price
            side: "long" or "short"
            funding_rate: Hourly funding rate
            risk_amount: Risk amount for % calculation
            hold_hours: Expected hold duration

        Returns:
            FundingEstimate with cost projections
        """
        notional = size * entry_price
        hourly_payment = notional * funding_rate

        if side == "long":
            projected_cost = hourly_payment * Decimal(str(hold_hours))
        else:
            projected_cost = -hourly_payment * Decimal(str(hold_hours))

        funding_eats_risk_pct = (
            max(Decimal("0"), projected_cost) / risk_amount
            if risk_amount > 0
            else Decimal("0")
        )

        return FundingEstimate(
            hourly_rate=funding_rate,
            hourly_cost=abs(hourly_payment) if projected_cost > 0 else Decimal("0"),
            hourly_income=abs(hourly_payment) if projected_cost < 0 else Decimal("0"),
            projected_24h=projected_cost,
            funding_eats_risk_pct=funding_eats_risk_pct,
        )

    def _round_down(self, value: Decimal, decimals: int) -> Decimal:
        """Round down to specified decimals (HL requirement)."""
        assert decimals >= 0, f"decimals must be >= 0, got {decimals}"

        if decimals == 0:
            return value.to_integral_value(rounding=ROUND_DOWN)
        quantize_str = "0." + "0" * decimals
        return value.quantize(Decimal(quantize_str), rounding=ROUND_DOWN)

    def _get_correlation_group(self, coin: str) -> str:
        """Get correlation group for a coin."""
        for group, coins in CORRELATION_MAP.items():
            if coin in coins:
                return group
        return f"independent-{coin}"

    def get_asset_id_from_meta(self, asset_meta: dict) -> int:
        """Extract asset ID from metadata."""
        return asset_meta.get("_asset_id", 0)

File: internal/risk/calculator.go
package risk

import (
	"fmt"

	"github.com/shopspring/decimal"

	"github.com/skilus/hyperhandler/internal/models"
)

// Candle is the minimal OHLC input the ATR calculation needs. The manager
// converts client candles into this shape so the calculator stays decoupled
// from the HTTP layer (and trivially testable). Mirrors the {h,l,c} dict fields
// used by calculator.py:calculate_atr.
type Candle struct {
	High  decimal.Decimal
	Low   decimal.Decimal
	Close decimal.Decimal
}

// Calculator holds the pure risk-calculation functions. Mirrors
// calculator.py:RiskCalculator.
type Calculator struct {
	Profile  RiskProfile
	HLConfig HLConfig
}

// NewCalculator builds a Calculator for the given profile and HL config.
func NewCalculator(profile RiskProfile, hlConfig HLConfig) *Calculator {
	return &Calculator{Profile: profile, HLConfig: hlConfig}
}

// CalculateATR computes ATR using the EMA method (pure decimal, no numpy).
// Returns an error if there are fewer than 2 candles. Mirrors
// calculator.py:calculate_atr.
func (c *Calculator) CalculateATR(candles []Candle, period int) (decimal.Decimal, error) {
	if len(candles) < 2 {
		return decimal.Zero, fmt.Errorf("Need at least 2 candles for ATR")
	}

	trueRanges := make([]decimal.Decimal, 0, len(candles)-1)
	for i := 1; i < len(candles); i++ {
		high := candles[i].High
		low := candles[i].Low
		prevClose := candles[i-1].Close

		tr := decimal.Max(
			high.Sub(low),
			high.Sub(prevClose).Abs(),
			low.Sub(prevClose).Abs(),
		)
		trueRanges = append(trueRanges, tr)
	}

	if len(trueRanges) < period {
		// Not enough data for full EMA, use simple average.
		sum := decimal.Zero
		for _, tr := range trueRanges {
			sum = sum.Add(tr)
		}
		return round28(sum.Div(decimal.NewFromInt(int64(len(trueRanges))))), nil
	}

	// EMA calculation: alpha = 2 / (period + 1). Python rounds every Decimal
	// operation to 28 significant figures (ROUND_HALF_EVEN); shopspring only
	// rounds on division, so we round each step explicitly to stay byte-equal.
	alpha := round28(decimal.NewFromInt(2).Div(decimal.NewFromInt(int64(period)).Add(decimal.NewFromInt(1))))
	oneMinusAlpha := round28(decimal.NewFromInt(1).Sub(alpha))
	ema := trueRanges[0]
	for _, tr := range trueRanges[1:] {
		term1 := round28(alpha.Mul(tr))
		term2 := round28(oneMinusAlpha.Mul(ema))
		ema = round28(term1.Add(term2))
	}
	return ema, nil
}

// round28 rounds to 28 significant figures with banker's rounding, emulating
// Python's default decimal context (prec=28, ROUND_HALF_EVEN). shopspring keeps
// full precision through Mul/Add/Sub, so sequential operations (notably the ATR
// EMA) must round at each step to match Python exactly. SPEC-007 risk #2.
func round28(d decimal.Decimal) decimal.Decimal {
	if d.IsZero() {
		return d
	}
	// adjExp is the exponent of the most-significant digit.
	adjExp := d.NumDigits() - 1 + int(d.Exponent())
	places := 27 - adjExp // decimal places that keep 28 significant figures
	return d.RoundBank(int32(places))
}

// CalculateStopLoss computes an ATR-based stop-loss price. Mirrors
// calculator.py:calculate_stop_loss.
func (c *Calculator) CalculateStopLoss(
	entryPrice decimal.Decimal,
	side string,
	atr decimal.Decimal,
	horizon models.SignalHorizon,
) models.StopLossResult {
	multiplier := ATRSettings[horizon].Multiplier

	stopDistance := atr.Mul(multiplier)
	buffer := stopDistance.Mul(c.HLConfig.SlippageBuffer)
	stopWithBuffer := stopDistance.Add(buffer)

	var stopPrice decimal.Decimal
	if side == "long" {
		stopPrice = entryPrice.Sub(stopWithBuffer)
	} else {
		stopPrice = entryPrice.Add(stopWithBuffer)
	}

	return models.StopLossResult{
		Price:         stopPrice,
		Distance:      stopWithBuffer,
		DistancePct:   stopWithBuffer.Div(entryPrice),
		ATRValue:      atr,
		ATRMultiplier: multiplier,
	}
}

// EstimateLiquidationPrice estimates the liquidation price for a NEW position
// (simplified HL cross-margin formula). Mirrors
// calculator.py:estimate_liquidation_price.
func (c *Calculator) EstimateLiquidationPrice(
	entryPrice decimal.Decimal,
	leverage int,
	side string,
) decimal.Decimal {
	liqDistancePct := decimal.NewFromInt(1).Div(decimal.NewFromInt(int64(leverage))).Sub(HLMaintenanceMargin)
	if floor := decimal.NewFromFloat(0.01); liqDistancePct.LessThan(floor) {
		liqDistancePct = floor // floor at 1%
	}

	if side == "long" {
		return entryPrice.Mul(decimal.NewFromInt(1).Sub(liqDistancePct))
	}
	return entryPrice.Mul(decimal.NewFromInt(1).Add(liqDistancePct))
}

// ValidateStopVsLiquidation reports whether the stop-loss sits closer to entry
// than liquidation, with the configured safety buffer. Mirrors
// calculator.py:validate_stop_vs_liquidation.
func (c *Calculator) ValidateStopVsLiquidation(
	stopPrice, liquidationPrice, entryPrice decimal.Decimal,
	side string,
) bool {
	if side == "long" {
		// For long: stop must be ABOVE liquidation.
		if stopPrice.LessThanOrEqual(liquidationPrice) {
			return false
		}
	} else {
		// For short: stop must be BELOW liquidation.
		if stopPrice.GreaterThanOrEqual(liquidationPrice) {
			return false
		}
	}

	// Check safety buffer (stop should be at least liq_safety_buffer away).
	liqBuffer := stopPrice.Sub(liquidationPrice).Abs().Div(entryPrice)
	return liqBuffer.GreaterThanOrEqual(c.HLConfig.LiqSafetyBuffer)
}

// SelectLeverage selects leverage so liquidation stays beyond the stop-loss.
// Mirrors calculator.py:select_leverage.
func (c *Calculator) SelectLeverage(
	stopDistancePct decimal.Decimal,
	maxLeverageCoin int,
) models.LeverageResult {
	safetyFactor := decimal.NewFromFloat(1.5)

	var maxSafe int
	if stopDistancePct.LessThanOrEqual(decimal.Zero) {
		maxSafe = c.Profile.MaxLeverage
	} else {
		maxSafe = int(decimal.NewFromInt(1).Div(stopDistancePct.Mul(safetyFactor)).IntPart())
		if maxSafe < 1 {
			maxSafe = 1
		}
	}

	leverage := minInt(maxSafe, maxLeverageCoin, c.Profile.MaxLeverage)

	var reason []string
	if leverage == maxSafe {
		reason = append(reason, "safe_for_stop")
	}
	if leverage == maxLeverageCoin {
		reason = append(reason, "coin_max")
	}
	if leverage == c.Profile.MaxLeverage {
		reason = append(reason, "config_max")
	}
	reasonStr := "default"
	if len(reason) > 0 {
		reasonStr = joinPlus(reason)
	}

	return models.LeverageResult{
		Leverage:  leverage,
		MaxSafe:   maxSafe,
		MaxCoin:   maxLeverageCoin,
		MaxConfig: c.Profile.MaxLeverage,
		Reason:    reasonStr,
	}
}

// SelectLeverageForStop selects leverage that makes the given stop valid (used
// when an ATR-based stop requires lower leverage). Mirrors
// calculator.py:select_leverage_for_stop.
func (c *Calculator) SelectLeverageForStop(
	stopPrice, entryPrice decimal.Decimal,
	side string,
	maxLeverageCoin int,
) models.LeverageResult {
	stopDistancePct := entryPrice.Sub(stopPrice).Abs().Div(entryPrice)

	// Liquidation must be further than stop + safety buffer.
	requiredLiqDistance := stopDistancePct.Add(c.HLConfig.LiqSafetyBuffer)

	// leverage ≈ 1 / (liq_distance + maintenance).
	maxSafe := int(decimal.NewFromInt(1).Div(requiredLiqDistance.Add(HLMaintenanceMargin)).IntPart())
	if maxSafe < 1 {
		maxSafe = 1
	}

	leverage := minInt(maxSafe, maxLeverageCoin, c.Profile.MaxLeverage)

	return models.LeverageResult{
		Leverage:  leverage,
		MaxSafe:   maxSafe,
		MaxCoin:   maxLeverageCoin,
		MaxConfig: c.Profile.MaxLeverage,
		Reason:    "adjusted_for_stop",
	}
}

// PositionSizeInput groups the arguments to CalculatePositionSize. Confidence
// is a pointer to mirror the Python `float | None` default.
type PositionSizeInput struct {
	AccountValue     decimal.Decimal
	AvailableBalance decimal.Decimal
	EntryPrice       decimal.Decimal
	StopPrice        decimal.Decimal
	Leverage         int
	SzDecimals       int
	Confidence       *float64
	RiskMultiplier   decimal.Decimal  // circuit-breaker multiplier (default 1.0)
	MaxRiskAmount    *decimal.Decimal // budget constraint, nil = unconstrained
}

// CalculatePositionSize sizes a position from the risk budget. Returns either a
// *models.PositionSizeResult or a *models.RiskReject (exactly one is non-nil),
// mirroring the Python union return of calculator.py:calculate_position_size.
func (c *Calculator) CalculatePositionSize(in PositionSizeInput) (*models.PositionSizeResult, *models.RiskReject) {
	// Clamp confidence factor.
	var confidenceFactor decimal.Decimal
	if in.Confidence != nil {
		confidenceFactor = decimal.NewFromFloat(*in.Confidence)
		if confidenceFactor.GreaterThan(MaxConfidenceFactor) {
			confidenceFactor = MaxConfidenceFactor
		}
		if confidenceFactor.LessThan(MinConfidenceFactor) {
			confidenceFactor = MinConfidenceFactor
		}
	} else {
		confidenceFactor = MaxConfidenceFactor
	}

	riskMultiplier := in.RiskMultiplier
	riskPct := c.Profile.RiskPerTrade.Mul(confidenceFactor).Mul(riskMultiplier)
	riskAmount := in.AccountValue.Mul(riskPct)

	// Apply budget constraint.
	if in.MaxRiskAmount != nil && riskAmount.GreaterThan(*in.MaxRiskAmount) {
		riskAmount = *in.MaxRiskAmount
	}

	stopDistance := in.EntryPrice.Sub(in.StopPrice).Abs()
	if stopDistance.IsZero() {
		return nil, &models.RiskReject{
			Reason:          models.RejectATRUnavailable,
			Details:         "Stop distance is zero",
			SuggestedAction: "wait",
		}
	}

	rawSize := riskAmount.Div(stopDistance)
	notional := rawSize.Mul(in.EntryPrice)
	commission := notional.Mul(c.HLConfig.TakerFee).Mul(decimal.NewFromInt(2))

	// Adjust for commission.
	adjustedRisk := riskAmount.Sub(commission)
	if adjustedRisk.LessThanOrEqual(decimal.Zero) {
		return nil, &models.RiskReject{
			Reason:          models.RejectPositionTooSmall,
			Details:         "Risk amount doesn't cover commission",
			SuggestedAction: "wait",
		}
	}

	adjustedSize := adjustedRisk.Div(stopDistance)

	// Check margin constraint.
	leverageDec := decimal.NewFromInt(int64(in.Leverage))
	marginRequired := adjustedSize.Mul(in.EntryPrice).Div(leverageDec)
	if marginRequired.GreaterThan(in.AvailableBalance) {
		maxSize := in.AvailableBalance.Mul(leverageDec).Div(in.EntryPrice)
		if maxSize.LessThan(adjustedSize) {
			adjustedSize = maxSize
		}
		marginRequired = adjustedSize.Mul(in.EntryPrice).Div(leverageDec)
	}

	// Round down to szDecimals.
	adjustedSize = roundDown(adjustedSize, in.SzDecimals)

	finalNotional := adjustedSize.Mul(in.EntryPrice)
	finalRisk := adjustedSize.Mul(stopDistance)
	finalCommission := finalNotional.Mul(c.HLConfig.TakerFee).Mul(decimal.NewFromInt(2))

	// Check minimum order.
	if finalNotional.LessThan(c.HLConfig.MinOrderValue) {
		return nil, &models.RiskReject{
			Reason:          models.RejectPositionTooSmall,
			Details:         fmt.Sprintf("Order $%s < min $%s", finalNotional.StringFixed(2), c.HLConfig.MinOrderValue.String()),
			SuggestedAction: "wait",
		}
	}

	riskPctOut := decimal.Zero
	if in.AccountValue.GreaterThan(decimal.Zero) {
		riskPctOut = finalRisk.Div(in.AccountValue)
	}

	return &models.PositionSizeResult{
		Size:               adjustedSize,
		Notional:           finalNotional,
		MarginRequired:     marginRequired,
		RiskAmount:         finalRisk,
		RiskPct:            riskPctOut,
		CommissionEstimate: finalCommission,
	}, nil
}

// CalculateCumulativeRisk computes portfolio cumulative risk with correlation
// adjustment and cascade buffer. Mirrors calculator.py:calculate_cumulative_risk.
func (c *Calculator) CalculateCumulativeRisk(
	openPositions []models.Position,
	newRiskAmount decimal.Decimal,
	newCoin string,
	accountValue decimal.Decimal,
) models.CumulativeRiskResult {
	// Build correlation groups from positions, preserving insertion order.
	groups := map[string][]string{}
	var groupOrder []string
	addGroup := func(name string) {
		if _, ok := groups[name]; !ok {
			groups[name] = []string{}
			groupOrder = append(groupOrder, name)
		}
	}
	for _, pos := range openPositions {
		group := c.correlationGroup(pos.Coin)
		addGroup(group)
		groups[group] = append(groups[group], pos.Coin)
	}

	// Add new coin if we're actually adding risk.
	if newRiskAmount.GreaterThan(decimal.Zero) {
		newGroup := c.correlationGroup(newCoin)
		addGroup(newGroup)
		if !containsStr(groups[newGroup], newCoin) {
			groups[newGroup] = append(groups[newGroup], newCoin)
		}
	}

	// Sum raw risk.
	totalRisk := decimal.Zero
	for _, pos := range openPositions {
		totalRisk = totalRisk.Add(posRisk(pos))
	}

	// Adjusted risk with correlation penalty + cascade buffer.
	adjustedRisk := decimal.Zero
	cascadeBuffer := decimal.Zero
	for _, name := range groupOrder {
		coins := groups[name]
		groupRisk := decimal.Zero
		for _, pos := range openPositions {
			if containsStr(coins, pos.Coin) {
				groupRisk = groupRisk.Add(posRisk(pos))
			}
		}

		n := len(coins)
		penalized := groupRisk
		if n > 1 {
			penalty := decimal.NewFromInt(1).Add(decimal.NewFromInt(int64(n - 1)).Mul(c.Profile.CorrelationFactor))
			penalized = groupRisk.Mul(penalty)
			cascadeBuffer = cascadeBuffer.Add(groupRisk.Mul(decimal.NewFromFloat(0.1)))
		}
		adjustedRisk = adjustedRisk.Add(penalized)
	}

	totalAdjusted := adjustedRisk.Add(cascadeBuffer).Add(newRiskAmount)
	maxRisk := accountValue.Mul(c.Profile.MaxCumulativeRisk)

	riskPct := decimal.Zero
	if accountValue.GreaterThan(decimal.Zero) {
		riskPct = totalAdjusted.Div(accountValue)
	}
	availableBudget := maxRisk.Sub(adjustedRisk).Sub(cascadeBuffer)
	if availableBudget.LessThan(decimal.Zero) {
		availableBudget = decimal.Zero
	}

	return models.CumulativeRiskResult{
		RawRisk:           totalRisk.Add(newRiskAmount),
		AdjustedRisk:      totalAdjusted,
		RiskPct:           riskPct,
		AvailableBudget:   availableBudget,
		WithinLimit:       totalAdjusted.LessThanOrEqual(maxRisk),
		CorrelationGroups: groups,
	}
}

// EstimateFundingCost estimates funding costs/income over the hold period.
// Mirrors calculator.py:estimate_funding_cost.
func (c *Calculator) EstimateFundingCost(
	size, entryPrice decimal.Decimal,
	side string,
	fundingRate, riskAmount decimal.Decimal,
	holdHours int,
) models.FundingEstimate {
	notional := size.Mul(entryPrice)
	hourlyPayment := notional.Mul(fundingRate)

	var projectedCost decimal.Decimal
	if side == "long" {
		projectedCost = hourlyPayment.Mul(decimal.NewFromInt(int64(holdHours)))
	} else {
		projectedCost = hourlyPayment.Neg().Mul(decimal.NewFromInt(int64(holdHours)))
	}

	fundingEats := decimal.Zero
	if riskAmount.GreaterThan(decimal.Zero) {
		fundingEats = decimal.Max(decimal.Zero, projectedCost).Div(riskAmount)
	}

	hourlyCost := decimal.Zero
	hourlyIncome := decimal.Zero
	if projectedCost.GreaterThan(decimal.Zero) {
		hourlyCost = hourlyPayment.Abs()
	} else if projectedCost.LessThan(decimal.Zero) {
		hourlyIncome = hourlyPayment.Abs()
	}

	return models.FundingEstimate{
		HourlyRate:         fundingRate,
		HourlyCost:         hourlyCost,
		HourlyIncome:       hourlyIncome,
		Projected24h:       projectedCost,
		FundingEatsRiskPct: fundingEats,
	}
}

// GetAssetIDFromMeta extracts the asset ID from metadata. Mirrors
// calculator.py:get_asset_id_from_meta; the index is threaded via the meta map
// under "_asset_id".
func (c *Calculator) GetAssetIDFromMeta(meta map[string]any) int {
	if v, ok := meta["_asset_id"]; ok {
		if id, ok := v.(int); ok {
			return id
		}
	}
	return 0
}

// correlationGroup returns the correlation group for a coin. Mirrors
// calculator.py:_get_correlation_group.
func (c *Calculator) correlationGroup(coin string) string {
	for group, coins := range CorrelationMap {
		if containsStr(coins, coin) {
			return group
		}
	}
	return "independent-" + coin
}

// roundDown rounds down (toward zero) to the given number of decimals. Mirrors
// calculator.py:_round_down (ROUND_DOWN). Panics on negative decimals to mirror
// the Python assertion.
func roundDown(value decimal.Decimal, decimals int) decimal.Decimal {
	if decimals < 0 {
		panic(fmt.Sprintf("decimals must be >= 0, got %d", decimals))
	}
	return value.Truncate(int32(decimals))
}

// posRisk returns a position's risk amount, treating nil as zero (mirrors the
// Python `pos.risk_amount or Decimal("0")`).
func posRisk(p models.Position) decimal.Decimal {
	if p.RiskAmount == nil {
		return decimal.Zero
	}
	return *p.RiskAmount
}

func containsStr(xs []string, target string) bool {
	for _, x := range xs {
		if x == target {
			return true
		}
	}
	return false
}

func minInt(xs ...int) int {
	m := xs[0]
	for _, x := range xs[1:] {
		if x < m {
			m = x
		}
	}
	return m
}

func joinPlus(parts []string) string {
	out := parts[0]
	for _, p := range parts[1:] {
		out += "+" + p
	}
	return out
}
